### ISSUE TITLE
Test suite for evaluating ABCI implementations for bounty

### ISSUE THREAD

**Issue by zramsay** [tendermint/tendermint#1848]
original issue: https://github.com/tendermint/abci/issues/76

TL:DR;

Here's what this still needs to be bounty ready:

- [ ] tests/server should have a struct that implements abci.Application so we can assert it supports all the ABCI methods
- [x] tests/server should not use fmt.Println. It should wrap all its errors (https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully)
- [ ] we should be running multiple distinct tests. It's fine for now if they run one after another and depend on the state of the previous one - let's just be clear about that in the code
  - [ ] BlockSequenceTest - run InitChain followed Info followed by a few instances of the block sequence (BeginBlock, multiple DeliverTx, EndBlock, Commit) with invalid deliver txs (so they dont affect the state). We don't really care what Info returns or what BeginBlock/EndBlock do here, but they shouldn't return errors (ie. the server implements them even if the Counter treats them as NOPs)
  - [ ] TxTest - run a bunch of DeliverTxs and CheckTxs where some are valid and some are invalid.  Make sure running CheckTx doesn't effect the results of DeliverTx, but DeliverTx effects the results of CheckTx.
  - [ ] CommitTest - run Commit. It should return the big-endian encoding of the txCount (ie. valid DeliverTxs sent so far). Run some more valid and invalid DeliverTx and run Commit again, make sure the result is correct.
  - [ ] QueryTest - run a Query for the "hash" and the "tx". "hash" should return the number of times we've called Commit, and "tx" the number of valid DeliverTx. Run some more DeliverTx and Commit, Query again, and confirm it was all updated right.

Note all CheckTx and DeliverTx should run via `Async`. Everything else can be `Sync`.

Also note that everything should be super neat and easy to see what's going on and easy to maintain, similar to how it looks now. 

Finally, we should clearly write the spec for the Counter (some of it is hidden in the above).

Then we can start rewarding people.

**Comment by jackzampolin** [comment]
Closing cause we aren't doing this bounty anymore.

**Comment by ebuchman** [comment]
I would love to keep this open

**Comment by marbar3778** [comment]
Closing cause we aren't doing this bounty anymore.

**Comment by KrzysiekJ** [comment]
So will bounty claims submitted when the bounty was still active be evaluated manually?

**Comment by marbar3778** [comment]
> So will bounty claims submitted when the bounty was still active be evaluated manually?

The team has changed a bit. Could you tell us where you submitted your work?

**Comment by KrzysiekJ** [comment]
Below is an email that I sent on `Tue, 04 Apr 2017 14:27:17 +0200` to `hello@tendermint.com` (subject: “ABCI server and counter in Erlang”). This was forwarded on `Fri, 28 Apr 2017 18:29:58 +0200` to `adrian@tendermint.com`.

The topic was also brought on the community channels and resulted in me submitting tendermint/abci#160 (see tendermint/abci#76 for some context).

As this ABCI server has been used in production for quite some time, I doubt that those automated tests are needed from the practical point of view in this case. Certainly they are not needed from the formal point of view (there was no such requirement).

> Hello,
>
> I’m happy to inform that I’ve released an ABCI server in Erlang:
>
> https://github.com/KrzysiekJ/abci_server
>
> There is an example counter application available:
>
> https://github.com/KrzysiekJ/abci_counter
>
> I’ve used the counter_test.sh file from the main Tendermint repository as a reference point for testing.
>
> My BTC bounty address is […].
>
> Recently I’ve also released an Erlang library which provides storage using general balanced binary Merkle trees:
>
> https://github.com/KrzysiekJ/gb_merkle_trees
>
> Regards,
> Krzysiek

I’ve redacted the BTC address, but can paste it in a comment here or send it in a DM.

**Comment by tessr** [comment]
@KrzysiekJ I will reach out to Tendermint Inc. to see what the status of this bounty is. (@marbar3778 and I are employees of Interchain GmbH, a separate company.)

**Comment by tessr** [comment]
@alessio will follow up on this 🙏🏼  thank you, Alessio!
